Cross framework serialization finish up

ViktorHofer opened the pull request:
This PR is intended to be the last piece to make serialization work across Desktop and Core.

## TODO
- [x] **mscorlib shim as default entrance point for BinaryFormatter** https://github.com/dotnet/corefx/pull/20879
- [x] **CookieContainer data storage (Hashtable vs Dictionary)** 
- [x] **Manual System.dll façade for TreeSet<T> and CookieVariant** https://github.com/dotnet/corefx/pull/20988
- [X] **Remaining manual equality checks in test code** [@krwq]
- [x] **OrdinalIgnoreCaseComparer does not exist on netfx** https://github.com/dotnet/coreclr/pull/12215 & https://github.com/dotnet/coreclr/pull/12267
- [x] **Fix Uri.OriginalString and uncomment equality check** [@krwq]
- [x] **Fix Cookie.TimeStamp and uncomment equality check** [@krwq]
- [x] **Waiting for https://github.com/dotnet/coreclr/pull/12267 to come into corefx**
## Status
Read from left to right. E.g.: _Core deserializes to Core_

|                     |       Core      |   Desktop     |
| ------------- | ------------- | ------------- |
| Core            |           x         |         x          |
| Desktop      |           x         |         x         |

justinvp commented:
> CookieContainer data storage (Hashtable vs Dictionary)

Note that the nested `CookieContainer.PathList` type is stored in the `CookieContainer`'s `Hashtable`/`Dictionary`... In the full framework, `PathList` uses the non-generic `SortedList`, but in CoreFX it uses the generic `SortedList<T>`.

ViktorHofer commented:
Thanks Justin. I already know that. We will discuss how to handle `CookieContainer` and all involved types later today and will keep you updated here.

danmoseley reviewed (approved):
LGTM

ViktorHofer commented:
@lmolkova how important is it that Activity is cross platform serializable? Do you need it to be interchangeable between Core and Desktop?

lmolkova commented:
I don't know how much context I miss.

I assume your question is related to cross AppDomain/process calls and presence of the Serializable attribute int the Activity. Note that Activity is serializable on NET 4.5 only.

The whle Activity is not intended to be passed outside of the process, only its certain properties should be propagated along with the request. E,g, if application makes HTTP call, some properties of the current Activity are passed in HTTP headers. So some protocol should decide how Activity properties are serialized in particual transport.

So Activity should not be interchangeable at all.

May be @vancem could also comment on that if I missed anything.

ViktorHofer commented:
@krwq types that probably need custom equality check:

- [ ] System.Collections.ArrayList
- [ ] System.Collections.BitArray
- [ ] System.Collections.Generic.ByteEqualityComparer
- [ ] System.Collections.Generic.Dictionary`2[System.Int32,System.String]
- [ ] System.Collections.Generic.EnumEqualityComparer`1[System.Runtime.Serialization.Formatters.Tests.Int32Enum]
- [ ] System.Collections.Generic.GenericComparer`1[System.Int32]
- [ ] System.Collections.Generic.GenericEqualityComparer`1[System.Int32]
- [ ] System.Collections.Generic.HashSet`1[System.Runtime.Serialization.Formatters.Tests.Point]
- [ ] System.Collections.Generic.LinkedList`1[System.Runtime.Serialization.Formatters.Tests.Point]
- [ ] System.Collections.Generic.List`1[System.Int32]
- [ ] System.Collections.Generic.NullableEqualityComparer`1[System.Double]
- [ ] System.Collections.Generic.ObjectEqualityComparer`1[System.Object]
- [ ] System.Collections.Generic.Queue`1[System.Int32]
- [ ] System.Collections.Generic.SortedList`2[System.Int32,System.Runtime.Serialization.Formatters.Tests.Point]
- [ ] System.Collections.Generic.SortedSet`1[System.Runtime.Serialization.Formatters.Tests.Point]
- [ ] System.Collections.Generic.Stack`1[System.Runtime.Serialization.Formatters.Tests.Point]
- [ ] System.Collections.Hashtable
- [ ] System.Collections.ObjectModel.Collection`1[System.Int32]
- [ ] System.Collections.ObjectModel.ObservableCollection`1[System.Int32]
- [ ] System.Collections.ObjectModel.ReadOnlyCollection`1[System.Int32]
- [ ] System.Collections.ObjectModel.ReadOnlyDictionary`2[System.Int32,System.String]
- [ ] System.Collections.ObjectModel.ReadOnlyObservableCollection`1[System.Int32]
- [ ] System.Collections.Queue
- [ ] System.Collections.SortedList
- [ ] System.Collections.Specialized.HybridDictionary
- [ ] System.Collections.Specialized.ListDictionary
- [ ] System.Collections.Specialized.NameValueCollection
- [ ] System.Collections.Specialized.OrderedDictionary
- [ ] System.Collections.Specialized.StringCollection
- [ ] System.Collections.Stack
- [ ] System.ComponentModel.BindingList`1[System.Int32]
- [ ] System.ComponentModel.BindingList`1[System.Runtime.Serialization.Formatters.Tests.Point]
- [ ] System.Data.PropertyCollection
- [ ] System.Globalization.CompareInfo
- [ ] System.Globalization.SortVersion
- [ ] System.Net.Cookie
- [ ] System.Net.CookieCollection
- [ ] System.Runtime.Serialization.Formatters.Tests.BasicISerializableObject
- [ ] System.Runtime.Serialization.Formatters.Tests.DerivedISerializableWithNonPublicDeserializationCtor
- [ ] System.Runtime.Serialization.Formatters.Tests.Graph`1[System.Int32]
- [ ] System.Runtime.Serialization.Formatters.Tests.ObjectWithArrays
- [ ] System.Runtime.Serialization.Formatters.Tests.ObjectWithIntStringUShortUIntULongAndCustomObjectFields
- [ ] System.Runtime.Serialization.Formatters.Tests.Point
- [ ] System.Runtime.Serialization.Formatters.Tests.Point[]
- [ ] System.Runtime.Serialization.Formatters.Tests.SealedObjectWithIntStringFields
- [ ] System.Runtime.Serialization.Formatters.Tests.SimpleKeyedCollection
- [ ] System.Runtime.Serialization.Formatters.Tests.Tree`1[System.Runtime.Serialization.Formatters.Tests.Colors]
- [ ] System.TimeZoneInfo
- [ ] System.Tuple`1[System.Int32]
- [ ] System.Tuple`2[System.Int32,System.String]
- [ ] System.Tuple`3[System.Int32,System.String,System.UInt32]
- [ ] System.Tuple`4[System.Int32,System.String,System.UInt32,System.Int64]
- [ ] System.Tuple`5[System.Int32,System.String,System.UInt32,System.Int64,System.Double]
- [ ] System.Tuple`6[System.Int32,System.String,System.UInt32,System.Int64,System.Double,System.Single]
- [ ] System.Tuple`7[System.Int32,System.String,System.UInt32,System.Int64,System.Double,System.Single,System.Decimal]
- [ ] System.Tuple`8[System.Int32,System.String,System.UInt32,System.Int64,System.Double,System.Single,System.Decimal,System.Tuple`1[System.Tuple`1[System.Int32]]]
- [ ] System.Uri
- [ ] System.Version

jkotas commented:
> custom equality check

What is it for?

ViktorHofer commented:
@lmolkova Thanks for the clarification. That means we don't to put TypeForwards on it and test roundtripping it from netfx to core and back.

ViktorHofer commented:
@jkotas to make sure that types are correctly serializing and deserializing from/to different platforms. Example given DataSet and DataTable don't override object Equals nor implement IEquatable<T>. Therefore we need to manually check the cloned and original objects serializable fields for equality.

krwq commented:
@ViktorHofer should we implement those in the product code?

jkotas commented:
> Example given DataSet and DataTable don't override object Equals nor implement IEquatable

Why is it a problem? When you serialize/deserialize object on full .NET Framework, Equals for master and clone sometimes returns true and sometimes false. That's by design. I do not think we need to be fixing it.

krwq commented:
@jkotas - I believe this was meant for tests only so that we are able to asses if the type is cross-round-tripping (netfx->corefx) correctly

jkotas commented:
I think I understand it now - it is just about making the test coverage better, not about fixing the product.

ViktorHofer commented:
> @ViktorHofer should we implement those in the product code?

If you want to go through API reviews for each type, then go for it. As this probably would take a lot of time and arguing and we want to test this till next week, I would not suggest it ;)

ViktorHofer commented:
> I think I understand it now - it is just about making the test coverage better, not about fixing the product.

Correct.

krwq committed:
wip: add stubs for custom isEqual

ViktorHofer committed:
Updated blobs, small code refactoring, disabled CookieContainer test

ViktorHofer committed:
Fixed netfx471 detection

krwq commented on src/System.Runtime.Serialization.Formatters/tests/EqualityExtensions.cs:
```diff
@@ -722,7 +722,7 @@ public static bool IsEqual(this SimpleKeyedCollection @this, SimpleKeyedCollecti
             if (!(@this != null &&
                 other != null &&
-                @this.Comparer.Equals(other.Comparer) &&
+                //@this.Comparer.Equals(other.Comparer) &&
```
?

ViktorHofer replied:
it was failing with it. I needed a clean test state.

ViktorHofer commented:
@morganbr OrdinalIgnoreCaseComparer does not exist in netfx. We changed this in Core which means that core blob --> netfx will fail as the type doesn't exist there.

Let's remove the serializable attribute from OrdinalIgnoreCaseComparer (file StringComparer.cs, coreclr).

krwq committed:
add custom equals for graph with cycles

krwq committed:
fix few more comparers

krwq commented:
@ViktorHofer Uri is failing currently. The diff is in Uri.OriginalString. Example failure:
```
Expected: https://dot.net
Actual:   https://dot.net/
```

krwq committed:
fix few more test cases

krwq committed:
fix few more types, fix GetExtensionMethod to work with parameters with same type name but different generic params

krwq commented:
@ViktorHofer Cookie class is broken and they differ by TimeStamp:
```
        Expected: 2017-06-12T14:15:31.3003955-07:00
        Actual:   2017-06-11T00:36:39.8488506
```

Likely we need to update test data and that is an issue for our test failures

krwq committed:
finish tests (2 comparisons are commented out: search for TODO)

krwq commented:
@ViktorHofer I think I'm done with tests

There are two cases which need to be uncommented and they are marked with TODO:
- [ ] Uri.OriginalString
- [ ] Cookie.TimeStamp

ViktorHofer commented:
@krwq Thanks, I will look at those two

ViktorHofer committed:
Merge

ViktorHofer commented:
The remaining errors deserializing Core blobs to Desktop:

1. new Cookie("host", "edge", "auth", "dot.net");
```
      System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests.ValidateAgainstBlobs(obj: host=edge, blobs: ["AAEAAAD/////AQAAAAAAAAAMAgAAADhTeXN0ZW0sIFZlcnNpb2"..., "AAEAAAD/////AQAAAAAA
  AAAMAgAAAElTeXN0ZW0sIFZlcnNpb2"...]) [FAIL]
        System.Runtime.Serialization.SerializationException : Unable to find assembly 'System.Net.Primitives, Version=4.1.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
```

2. new ObjectWithArrays
            {
                IntArray = new int[0],
                StringArray = new string[] { "hello", "world" },
                ByteArray = new byte[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19 },
                JaggedArray = new int[][] { new int[] { 1, 2, 3 }, new int[] { 4, 5, 6, 7 } },
                MultiDimensionalArray = new int[,] { { 1, 2 }, { 3, 4 }, { 5, 6 } },
                TreeArray = new Tree<int>[] { new Tree<int>(1, new Tree<int>(2, null, null), new Tree<int>(3, null, null)) }
            }
```
System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests.ValidateAgainstBlobs(obj: ObjectWithArrays { ByteArray = [0, 1, 2, 3, 4, ...], IntArray = [], JaggedArray = [[...], [...]],
   MultiDimensionalArray = [1, 2, 3, 4, 5, ...], StringArray = ["hello", "world"], ... }, blobs: ["AAEAAAD/////AQAAAAAAAAAMAgAAAHBTeXN0ZW0uUnVudGltZS"..., "AAEAAAD/////AQAAAAAAAAAMAgAAAHBTeXN0ZW0u
  UnVudGltZS"...]) [FAIL]
        System.Runtime.Serialization.SerializationException : Unable to find assembly 'System.Runtime.Serialization.Formatters.Tests, Version=4.0.2.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b6
  8eb'.
```

3. new System.Collections.Specialized.ListDictionary(StringComparer.Create(CultureInfo.InvariantCulture, false))
            {
                { "key1", "value1" },
                { "keyx", "valuex" }
            };
```
      System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests.ValidateAgainstBlobs(obj: [System.Collections.DictionaryEntry, System.Collections.DictionaryEntry], blobs: ["AAEAAAD/////AQ
  AAAAAAAAAMAgAAAElTeXN0ZW0sIFZlcnNpb2"..., "AAEAAAD/////AQAAAAAAAAAMAgAAAElTeXN0ZW0sIFZlcnNpb2"...]) [FAIL]
        System.Runtime.Serialization.SerializationException : Unable to find assembly 'System.Collections.Specialized, Version=4.1.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
```

4. new System.Collections.Specialized.ListDictionary(StringComparer.OrdinalIgnoreCase)
```
      System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests.ValidateAgainstBlobs(obj: [], blobs: ["AAEAAAD/////AQAAAAAAAAAMAgAAAElTeXN0ZW0sIFZlcnNpb2"..., "AAEAAAD/////AQAAAAAAAAAMAgAAAElTeXN0ZW0sIFZlcnNpb2"...]) [FAIL]
        System.Runtime.Serialization.SerializationException : Unable to load type System.OrdinalIgnoreCaseComparer required for deserialization.
```

5. new System.Collections.Specialized.HybridDictionary(10, true)
            {
                { 32, "5d" },
                { "5d", 32 }
            };
```
      System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests.ValidateAgainstBlobs(obj: [System.Collections.DictionaryEntry, System.Collections.DictionaryEntry], blobs: ["AAEAAAD/////AQ
  AAAAAAAAAMAgAAAElTeXN0ZW0sIFZlcnNpb2"..., "AAEAAAD/////AQAAAAAAAAAMAgAAAElTeXN0ZW0sIFZlcnNpb2"...]) [FAIL]
        System.InvalidCastException : Object must implement IConvertible.
```

ViktorHofer committed:
Adding TypeForward for DictionaryNode

ViktorHofer commented on src/System.Collections.Specialized/src/System/Collections/Specialized/ListDictionary.cs:
```diff
@@ -493,6 +492,7 @@ public void Reset()
         }
 
         [Serializable]
+        [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
```
@weshaggard @jkotas Without this TypeForwardedFrom attribute for this nested type I would get a blob which contains the assembly information of System.Collection.Spezialized, Version=... when serializing it in Core

```
�����������
���ISystem, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
���aSystem.Collections.Specialized, Version=4.1.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a���-System.Collections.Specialized.ListDictionary���headversioncountcomparer��<System.Collections.Specialized.ListDictionary+DictionaryNode���System.CultureAwareComparer���	���������	������<System.Collections.Specialized.ListDictionary+DictionaryNode���keyvaluenext<System.Collections.Specialized.ListDictionary+DictionaryNode���������keyx���valuex
���System.CultureAwareComparer���
_compareInfo
_ignoreCase� System.Globalization.CompareInfo	������� System.Globalization.CompareInfo���m_name
m_sortVersion System.Globalization.SortVersion	����
```

With my change:
```
�����������
���ISystem, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089���-System.Collections.Specialized.ListDictionary���headversioncountcomparer��<System.Collections.Specialized.ListDictionary+DictionaryNode���System.CultureAwareComparer���	���������	������<System.Collections.Specialized.ListDictionary+DictionaryNode���keyvaluenext<System.Collections.Specialized.ListDictionary+DictionaryNode���������keyx���valuex
���System.CultureAwareComparer���
_compareInfo
_ignoreCase� System.Globalization.CompareInfo	������� System.Globalization.CompareInfo���m_name
m_sortVersion System.Globalization.SortVersion����
```

As you told me that we don't need type forwards on nested types am curious why it isn't working here without it?

jkotas replied:
Maybe we were wrong and the `TypeForwardedFrom` attributes are needed even on nested types.

weshaggard replied:
I guess the only real way to know is to debug through the BinaryFormatter code and see if it looks at the TypeForwardedFrom attributes on nested types.

ViktorHofer replied:
They are at least needed on types which live in mscorlib on netfx and are not in System.Private.Corelib on core for the optimization of BinaryFormatter (strip assembly full names).

ViktorHofer committed:
Updating blobs

ViktorHofer committed:
Update Cookie hashes, adding TypeForwards, removing unnecessary equality checks for CookieCollection

ViktorHofer commented on src/System.Net.Primitives/src/System/Net/Cookie.cs:
```diff
@@ -10,6 +10,7 @@
 
 namespace System.Net
 {
+    [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, PublicKeyToken=b77a5c561934e089")]
```
@weshaggard this is now another reason why we need a manual System.dll shim.

weshaggard replied:
This one wouldn't require the manual System.dll shim as it is public so it should already be there.

ViktorHofer replied:
Does it also need to be exposed in the reference assemblies for that? Because it isn't exposed there...

weshaggard replied:
Oh it is internal on .NET Framework so yeah you need to the manual shim for this one.

ViktorHofer committed:
Reenabling SortedDictionary test

ViktorHofer commented:
@krwq if you have free cycles, can you take a look at those two and try to fix them?

krwq committed:
make cookie.Timestamp to roundtrip correctly

krwq committed:
regenerate netfx and corefx blobs to account for Cookie.Timestamp change

ViktorHofer committed:
Update blobs

krwq committed:
make regex fix the blob for ObjectWithArray

krwq committed:
fix ObjectWithArray netfx test failure

krwq commented:
@ViktorHofer I haven't started on CookieContainer yet so you can pick it up in the morning if that will be the last thing, if not I'll do it tomorrow.

krwq commented:
@krwq-bot show failures

ViktorHofer committed:
Replacing Dictionary with Hashtable for desktop serialization support

ViktorHofer committed:
Update blobs

ViktorHofer committed:
Fixing a lot of TypeForwards, adding another hashtable test, other fixes

ViktorHofer committed:
Adding a few more TypeForwardedFrom attributes

ViktorHofer committed:
Updating the coreclr version, disabling CookieContainer test

ViktorHofer committed:
Update blobs

ViktorHofer commented:
@stephetoub @danmosemsft This is now ready to merge and will unblock the maestro PR from coreclr. Lots of changes since the last review... Please take the time and verify them.

danmosemsft commented:
@stephentoub can you help review?

ViktorHofer commented:
@stephentoub @danmosemsft after you reviewed it, if there are no changes required please merge it and trigger the tests here again: https://github.com/dotnet/corefx/pull/21012

ViktorHofer commented:
Failing test gets fixed by this: https://github.com/dotnet/coreclr/pull/12273

danmosemsft commented:
@morganbr if you have cycles, perhaps you could take a look at changes since your last review, as we believe this is done now (except for more tests)

ViktorHofer committed:
Disabling CompareInfo test --> waiting for coreclr

ViktorHofer commented:
I disabled the CompareInfo test. After CI is green and somebody signed of this should be mergeable. I will reactivate the test in the next PR

danmosemsft commented:
@davidsh or @CIPop could you please look at the CookieContainer changes?

danmosemsft commented on src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterHelper.cs:
```diff
@@ -16,16 +17,29 @@ public partial class BinaryFormatterTests
+            }
+            catch (Exception)
+            {
+                Console.WriteLine("Error during equality check of type " + obj?.GetType()?.FullName);
```
Not necessary to change now, but it's better to use Assertion rather than doing Console.Writeline as the Console output is not associated with the test (and is not in the test log)

In this case it would basically be Assertion.IsTrue(false, ... string here...)

In the case where you want to add a message to a thrown AssertionException, you would throw a new AssertionException including the callstack of the old one.

krwq replied:
@danmosemsft this is wrapping assertion because Assert gives really not readable message

morganbr replied:
Can't you just add a message to the assert?

danmosemsft replied:
Assert.True takes a message.

ViktorHofer replied:
done

danmosemsft commented on src/System.Runtime.Serialization.Formatters/tests/EqualityExtensions.cs:
```diff
@@ -130,5 +135,759 @@ public static bool IsEqual(this StringDictionary @this, StringDictionary other)
+                @this.Query == other.Query &&
+                @this.Fragment == other.Fragment &&
+                @this.Scheme == other.Scheme &&
+                //@this.OriginalString == other.OriginalString && // TODO
```
@krwq is this still TODO

krwq replied:
@ViktorHofer have you reverted some of my changes? I have removed this line - we do not serialize this field - @ViktorHofer please check if you haven't reverted more things when reverting

ViktorHofer replied:
done

danmosemsft commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -810,22 +815,24 @@ private void BuildCookieCollectionFromDomainMatches(Uri uri, bool isSecure, int
                 PathList pathList;
                 lock (m_domainTable)
```
Oops, this should be `lock (m_domainTable.SyncRoot)`

danmosemsft replied:
DEfinitely should fix that.

danmosemsft replied:
Same issue on line 866 below the comment `// (This is the only place that does domain removal)`

ViktorHofer replied:
done

danmosemsft commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -810,22 +815,24 @@ private void BuildCookieCollectionFromDomainMatches(Uri uri, bool isSecure, int
                 lock (m_domainTable)
                 {
-                    if (!m_domainTable.TryGetValue(domainAttribute[i], out pathList))
+                    object pathListValue = m_domainTable[domainAttribute[i]];
```
There is no need for `pathListValue ` you can just assign directly to `pathList`

danmosemsft replied:
Looks like desktop does that.

ViktorHofer replied:
not changing

danmosemsft commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -950,7 +958,7 @@ internal string GetCookieHeader(Uri uri, out string optCookie2)
             string delimiter = string.Empty;
 
-            var builder = StringBuilderCache.Acquire();
+            StringBuilder builder = StringBuilderCache.Acquire();
```
FYI for ask-mode changes, we try to avoid unnecessary unrelated changes like this. Reasons are (1) distracting from actual diff (2) in general, adds risk. Not worth changing this now, especially because of the volume of other changes...

ViktorHofer replied:
ok

ViktorHofer commented:
pathList field is defined as PathList which is a struct which doesn't support null checks

Sent from my Windows 10 phone

From: Dan Moseley<mailto:notifications@github.com>
Sent: Wednesday, June 14, 2017 8:31 PM
To: dotnet/corefx<mailto:corefx@noreply.github.com>
Cc: Viktor Hofer<mailto:Viktor.Hofer@microsoft.com>; Mention<mailto:mention@noreply.github.com>
Subject: Re: [dotnet/corefx] Cross framework serialization finish up (#20870)


@danmosemsft commented on this pull request.

________________________________

In src/System.Net.Primitives/src/System/Net/CookieContainer.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcorefx%2Fpull%2F20870%23discussion_r122029066&data=02%7C01%7Cviktor.hofer%40microsoft.com%7Cd1bb24da59204f6f31d208d4b353a013%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636330619129786744&sdata=zh1PzUrpmFM6EO7%2Bz0vuXMiTmnlJzTHhyrs0Mh0thLs%3D&reserved=0>:

> @@ -810,22 +815,24 @@ private void BuildCookieCollectionFromDomainMatches(Uri uri, bool isSecure, int
                 PathList pathList;
                 lock (m_domainTable)
                 {
-                    if (!m_domainTable.TryGetValue(domainAttribute[i], out pathList))
+                    object pathListValue = m_domainTable[domainAttribute[i]];


There is no need for pathListValue you can just assign directly to pathList

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcorefx%2Fpull%2F20870%23pullrequestreview-44108301&data=02%7C01%7Cviktor.hofer%40microsoft.com%7Cd1bb24da59204f6f31d208d4b353a013%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636330619129786744&sdata=eIWb4c7j3%2Bp%2FcQ589NbHNYEQGn%2Fi2S7QW4VDZA0lRUo%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAHEbq0bmSYn6zMMAuUo5rWUvuZ6eTJyWks5sECcUgaJpZM4N1eOV&data=02%7C01%7Cviktor.hofer%40microsoft.com%7Cd1bb24da59204f6f31d208d4b353a013%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636330619129786744&sdata=S%2FlKLq8zlxfQrEeYPqgf30l7QCK9bArJ0%2BKEqRGi%2Bnk%3D&reserved=0>.

danmosemsft commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -1013,24 +1019,29 @@ public int GetCookiesCount()
         }
 
-        public CookieCollection this[string s]
+        public ICollection Values
```
do we need to add this?

danmosemsft replied:
Although I see desktop had it (and with no explicit lock) so I guess it is best to add it as you did.

ViktorHofer replied:
ok

danmosemsft commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -983,15 +991,16 @@ public void SetCookies(Uri uri, string cookieHeader)
+        private readonly SortedList m_list; // Do not rename (binary serialization)
 
-        public static PathList Create() => new PathList(new SortedList<string, CookieCollection>(PathListComparer.StaticInstance));
+        public static PathList Create() => new PathList(SortedList.Synchronized(new SortedList(PathListComparer.StaticInstance)));
```
I'm not sure why this was `.Sychronized` before since evidently when we changed to the generic form we concluded that the locks below were enough. Especially since we exposed no way to write to it.

But, it is safest to match desktop here as you have done.

morganbr replied:
I'm not sure whether we decided to make synchronized SortedLists serializable. I'd suggest checking.

ViktorHofer replied:
I know. That's why I changed it: https://github.com/dotnet/corefx/pull/20870/files#diff-f48fff7b7909f489c5e8b8599c22843cR614

ViktorHofer commented:
It’s also exposed in netfx. I guess we could workaround it.

From: Dan Moseley<mailto:notifications@github.com>
Sent: Wednesday, June 14, 2017 8:37 PM
To: dotnet/corefx<mailto:corefx@noreply.github.com>
Cc: Viktor Hofer<mailto:Viktor.Hofer@microsoft.com>; Mention<mailto:mention@noreply.github.com>
Subject: Re: [dotnet/corefx] Cross framework serialization finish up (#20870)


@danmosemsft commented on this pull request.

________________________________

In src/System.Net.Primitives/src/System/Net/CookieContainer.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcorefx%2Fpull%2F20870%23discussion_r122030548&data=02%7C01%7Cviktor.hofer%40microsoft.com%7Cdabec24321a04ba1031b08d4b354627b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636330622390279405&sdata=ZkMh1zsgFmUKxBRK2mwYk8oxIbhRe6z%2Bn4Vxw%2BlJ3H0%3D&reserved=0>:

>                      count += cc.Count;
                 }
             }
             return count;
         }

-        public CookieCollection this[string s]
+        public ICollection Values


do we need to add this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcorefx%2Fpull%2F20870%23pullrequestreview-44109902&data=02%7C01%7Cviktor.hofer%40microsoft.com%7Cdabec24321a04ba1031b08d4b354627b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636330622390279405&sdata=8oAEa7mAIga8WnwRqbIDwJfMxY6jIaQK8OsbenWSLQU%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAHEbq1W4Ju_BbNdP9HCkhhtvnV1XciFfks5sEChbgaJpZM4N1eOV&data=02%7C01%7Cviktor.hofer%40microsoft.com%7Cdabec24321a04ba1031b08d4b354627b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636330622390279405&sdata=wnIo%2BaYZUZpK%2FeTOAJuXTtiaPWL9WiebsLAb1Z%2FAfqA%3D&reserved=0>.

davidsh commented on src/System.Net.Primitives/src/System/Net/Cookie.cs:
```diff
@@ -48,15 +49,16 @@ public sealed class Cookie
         private bool m_discard = false; // Do not rename (binary serialization)
         private string m_domain = string.Empty; // Do not rename (binary serialization)
-        private bool m_domainImplicit = true; // Do not rename (binary serialization)
+        private bool m_domain_implicit = true; // Do not rename (binary serialization)
```
Why are these variables being renamed to use underscores?

danmosemsft replied:
Must match desktop to serialize between.

davidsh commented on src/System.Net.Primitives/src/System/Net/Cookie.cs:
```diff
@@ -590,14 +592,13 @@ public string Port
                         if (ports[i] != string.Empty)
                         {
-                            if (!Int32.TryParse(ports[i], out port))
+                            if (!Int32.TryParse(ports[i], out int port))
```
what is the purpose of this change?

danmosemsft replied:
This looks like cleanup, which we can revert.

ViktorHofer replied:
I reverted it

davidsh commented on src/System.Net.Primitives/src/System/Net/CookieCollection.cs:
```diff
@@ -11,7 +11,7 @@ namespace System.Net
     // A list of cookies maintained in Sorted order. Only one cookie with matching Name/Domain/Path
     [Serializable]
-    [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, PublicKeyToken=b77a5c561934e089")]
+    [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
```
Why are we adding "Culture=neutral" attributes?  Is this different from .NET Framework?

ViktorHofer replied:
Because we need it for BinaryFormatter blob optimization. If you want to know it in more detail just ping me

davidsh commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -100,8 +100,7 @@ public class CookieContainer
 
-        // NOTE: all accesses of _domainTable must be performed with _domainTable locked.
-        private readonly Dictionary<string, PathList> m_domainTable = new Dictionary<string, PathList>(); // Do not rename (binary serialization)
+        private readonly Hashtable m_domainTable = new Hashtable(); // Do not rename (binary serialization)
```
Why change to Hashtable?

danmosemsft replied:
Because Desktop uses Hashtable and we are using vanilla binary serialization so the types and names must match.

danmosemsft commented:
> pathList field is defined as PathList which is a struct which doesn't support null checks

Ah yes I didn't see that. Fine.

krwq committed:
generate test cases

justinvp commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -381,27 +385,26 @@ private bool AgeCookies(string domain)
+            lock (m_domainTable.SyncRoot)
             {
-                foreach (KeyValuePair<string, PathList> entry in m_domainTable)
+                foreach (DictionaryEntry entry in m_domainTable)
```
This is going to allocate a boxed object for each entry in the dictionary (see #7539, #9050, #10247). It'd be better to do something like:

```c#
// Manual use of IDictionaryEnumerator instead of foreach to avoid DictionaryEntry box allocations.
IDictionaryEnumerator e = m_domainTable.GetEnumerator();
while (e.MoveNext())
{
    if (domain == null)
    {
        tempDomain = (string)e.Key;
        pathList = (PathList)e.Value;
    }
    ...
}
```

ViktorHofer replied:
> We can change that later. Or even better, if you want to help you could spin up a PR which optimizes my recent changes without killing the just introduced serialization support.

justinvp commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -810,22 +815,24 @@ private void BuildCookieCollectionFromDomainMatches(Uri uri, bool isSecure, int
                 lock (pathList.SyncRoot)
                 {
-                    foreach (KeyValuePair<string, CookieCollection> pair in pathList)
+                    foreach (DictionaryEntry entry in pathList)
```
This is going to allocate a boxed object for each entry in the dictionary. I wonder if `PathList.GetEnumerator` should be changed to return `IDictionaryEnumerator` instead of `IEnumerator`, so the allocations can be avoided while iterating through pathList.

ViktorHofer replied:
> We can change that later. Or even better, if you want to help you could spin up a PR which optimizes my recent changes without killing the just introduced serialization support.

justinvp commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -1061,14 +1072,15 @@ public object SyncRoot
         [Serializable]
-        private sealed class PathListComparer : IComparer<string>
+        [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
+        sealed class PathListComparer : IComparer
```
`private`

ViktorHofer replied:
I thought this was needed for cross framework serialization. I might be wrong

justinvp commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -983,15 +991,16 @@ public void SetCookies(Uri uri, string cookieHeader)
 
     [Serializable]
+    [System.Runtime.CompilerServices.TypeForwardedFrom("System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
     internal struct PathList
```
Now that this is going back to being stored in `Hashtable` (and therefore is going to be boxed anyway), should it be changed back to a class?

ViktorHofer replied:
We can change that later. Or even better, if you want to help you could spin up a PR which optimizes my recent changes without killing the just introduced serialization support.

krwq committed:
update blobs, comment out not working cases

krwq committed:
fix null equality checks for new test cases and reenable them

morganbr commented on src/System.Net.Primitives/src/System/Net/CookieContainer.cs:
```diff
@@ -737,8 +742,8 @@ internal CookieCollection InternalGetCookies(Uri uri)
 
-            List<string> domainAttributeMatchAnyCookieVariant = new List<string>();
-            List<string> domainAttributeMatchOnlyCookieVariantPlain = null;
+            var domainAttributeMatchAnyCookieVariant = new Collections.Generic.List<string>();
```
Using a partial namespace here is a bit strange looking. Why not just leave it as List<string>? If we need namespace qualification, please use the full namespace.

ViktorHofer replied:
The compiler couldn't distinguish between SortedList generic and non-generic. I will add the fully namespace here.

ViktorHofer replied:
done

morganbr commented on src/System.Runtime.Serialization.Formatters/tests/EqualityExtensions.cs:
```diff
@@ -130,5 +135,759 @@ public static bool IsEqual(this StringDictionary @this, StringDictionary other)
+                BinaryFormatterTests.CheckEquals(@this.Comparer, other.Comparer)))
+                return false;
+
+            return BinaryFormatterTests.CheckSequenceEquals(@this, other);
```
I'm not sure ordering is guaranteed in different hash sets

ViktorHofer replied:
@krwq please review this comment

krwq replied:
@morganbr @ViktorHofer currently no failures related to that so seems like order is preserved

morganbr replied:
@krwq , I expect that ordering may change in the future, such as when #14354 is completed. It would be good if this test wasn't sensitive to that.

morganbr commented on src/System.Runtime.Serialization.Formatters/tests/EqualityExtensions.cs:
```diff
@@ -130,5 +135,759 @@ public static bool IsEqual(this StringDictionary @this, StringDictionary other)
+                @this.IsReadOnly == other.IsReadOnly &&
+                @this.IsFixedSize == other.IsFixedSize &&
+                @this.IsSynchronized == other.IsSynchronized &&
+                BinaryFormatterTests.CheckSequenceEquals(@this.Keys, other.Keys) &&
```
Same set ordering issue

morganbr commented on src/System.Runtime.Serialization.Formatters/tests/EqualityExtensions.cs:
```diff
@@ -130,5 +135,759 @@ public static bool IsEqual(this StringDictionary @this, StringDictionary other)
+                other != null &&
+                BinaryFormatterTests.CheckEquals(@this.Comparer, other.Comparer) &&
+                @this.Count == other.Count &&
+                BinaryFormatterTests.CheckSequenceEquals(@this.Keys, other.Keys) &&
```
Same set ordering issue

ViktorHofer replied:
@krwq please review this comment

krwq replied:
@morganbr @ViktorHofer because we don't have any failures related to that I'd leave as is - I wouldn't fix more than needed in this PR

morganbr commented on src/System.Runtime.Serialization.Formatters/tests/EqualityExtensions.cs:
```diff
@@ -130,5 +135,759 @@ public static bool IsEqual(this StringDictionary @this, StringDictionary other)
+        {
+            if (!(@this != null &&
+                other != null &&
+                BinaryFormatterTests.CheckSequenceEquals(@this.Keys, other.Keys) &&
```
Same set ordering issue

ViktorHofer replied:
@krwq please review this comment

morganbr commented on src/System.Runtime.Serialization.Formatters/tests/EqualityExtensions.cs:
```diff
@@ -130,5 +135,759 @@ public static bool IsEqual(this StringDictionary @this, StringDictionary other)
+            if (!(@this != null &&
+                other != null &&
+                @this.Count == other.Count &&
+                BinaryFormatterTests.CheckSequenceEquals(@this.Keys, other.Keys) &&
```
Another possible set ordering issue

ViktorHofer replied:
@krwq please review this comment

ViktorHofer committed:
PR feedback addressed

ViktorHofer committed:
Merge branch 'SerilalizationFinishUp' of https://github.com/ViktorHofer/corefx into SerilalizationFinishUp

ViktorHofer committed:
Fixed a typo...

danmosemsft commented on src/System.Runtime.Serialization.Formatters/tests/EqualityExtensions.cs:
```diff
@@ -130,5 +135,764 @@ public static bool IsEqual(this StringDictionary @this, StringDictionary other)
+
+        public static bool IsEqual(this LinkedList<Point> @this, LinkedList<Point> other)
+        {
+            if (!(@this != null &&
```
@krwq many of these IsEqual implementations are missing this I thikn?
```C#
            if (@this == null && other == null)
               return true;
```

danmosemsft replied:
Ah, that's handled in CheckSequenceEquals. fine.

krwq replied:
I'm only fixing when I understand why the value is null, in many cases it should not be null even though that would be correct to check it in here

danmosemsft commented:
All LGTM as of now...

ViktorHofer committed:
Updating used coreclr version, enabling compareinfo test again
